### PR TITLE
Event subsystem

### PR DESCRIPTION
**Pull request by Doralitze**
This system makes external events visible to the filter processing.

### CODE REVIEW

File: src/events/event_source.cpp [hunk 1]
```diff
@@ -0,0 +1,14 @@
...
+
+#include "event_source.hpp"
+#include "event_storage.hpp"
+
+namespace dmxfish::events {
+    event_source::event_source() : sender_id{} {}
+
+    void event_source::deregister() {
+
+    }
```
Comment (LRahlff) [inline src/events/event_source.cpp:13]: is it right that this function does nothing?

**Comment by LRahlff** [comment]
what are these events? input events? and who does handle the event, or is it just an information point, so a filter can call get_storage, to see all events actual working on. And because swap buffers gets called once a cycle, the filters can read also the single_trigger events, right?

**Review by LRahlff — APPROVED** [review]

File: src/events/event_source.cpp (hunk 1 above)
Comment (Doralitze) [reply src/events/event_source.cpp:13]: Yes, it hasn't been implemented yet. I forgot to do so and the unit test apparently lacks a proper check for this.

**Comment by Doralitze** [comment]
> what are these events? input events?

For example the input from the macro keyboard or incoming midi control sequences or remote tcp control commands.

> and who does handle the event, or is it just an information point, so a filter can call get_storage, to see all events actual working on. And because swap buffers gets called once a cycle, the filters can read also the single_trigger events, right?

Each filter can iterate through these events and check if something relevant happed for it or not. For example the sequence stack (#12) would scan these events in order to trigger sequences.

File: src/events/event_storage.cpp [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+                    break;
+                case event_type::RELEASE:
+                    this->ongoing_events.erase(ev.get_event_sender());
+                    break;
+            }
+        }
+
+        for (auto& lasting_ev : this->ongoing_events) {
+            new_read_storage.push_back(lasting_ev.second);
+        }
```
Comment (LRahlff) [inline src/events/event_storage.cpp:61]: for me it seems a little bit weird, that a event (first time START) is twice in the storage, once with START and once with ONGOING_EVENT.
Is it right that an event started and released in one storage does not get recognized as ongoing_event?
I think it is good, but could do strange things in the filter (if it should get recognized, but one is only looking for the ongoing events)

Comment (Doralitze) [reply src/events/event_storage.cpp:61]: Well, I kinda did that on purpose, as filter might either look for a rising edge or for the continuous presence but usually not for both. In the rare case that they do, they need to they need to implement proper precautions themself.

I also did not implement a lookup index for events as our event objects are small and presumably few. As a consequence I expect linear iteration to be faster than building a hash map every iteration and then performing lookups on them but we need to profile this in the real world.

**Pull request merged**